### ISSUE TITLE
Installing VisualStudio appears to work, but actually doesn't

### ISSUE THREAD

**Issue by TemporarilyOffline** [chocolatey/chocolatey-cookbook#52]
Not sure if this is the right place to put this...

I have a chef recipe that runs the following code block:

```ruby
include_recipe 'chocolatey'

chocolatey 'visualstudio2013ultimate' 

chocolatey 'visualstudio2013ultimate' do
  args '--force'
end
```
On the first run I get the output:

```
       Recipe: windows-build-agent::chocostudio
       
           * execute[install package visualstudio2013ultimate] action run
             - execute "C:\ProgramData\chocolatey/bin/choco" install -y  visualstudio2013ultimate
         
        (up to date)
```

On the second run I get the output:

```
       Recipe: windows-build-agent::chocostudio
        (up to date)
        (up to date)
```
neither of them actually install VS.  If I log into the instance and run choco install, I get this:

```
PS C:\programdata\chocolatey> .\choco.exe install visualstudio2013ultimate
Chocolatey v0.9.9.8
Installing the following packages:
visualstudio2013ultimate
By installing you accept licenses for the packages.
VisualStudio2013Ultimate v12.0.21005.20141031 already installed.
 Use --force to reinstall, specify a version to install, or try upgrade.

Chocolatey installed 0/1 package(s). 0 package(s) failed.
 1 package(s) had warnings.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
Warnings:
 - visualstudio2013ultimate
```
The programs do not show up on disk, nor start menu.  The log file is as follows:

```
2015-09-23 18:22:13,475 [DEBUG] - Command line: "C:\ProgramData\chocolatey\choco.exe" install -y visualstudio2013ultimate
2015-09-23 18:22:13,475 [DEBUG] - Received arguments: install -y visualstudio2013ultimate
2015-09-23 18:22:13,553 [DEBUG] - 
NOTE: Hiding sensitive configuration data! Please double and triple 
 check to be sure no sensitive data is shown, especially if copying 
 output to a gist for review.
2015-09-23 18:22:13,569 [DEBUG] - Configuration: CommandName='install'|
CacheLocation='C:\Users\ADMINI~1\AppData\Local\Temp'|
ContainsLegacyPackageInstalls='True'|
CommandExecutionTimeoutSeconds='2700'|
Sources='https://chocolatey.org/api/v2/'|Debug='False'|Verbose='False'|
Force='False'|Noop='False'|HelpRequested='False'|RegularOutput='True'|
PromptForConfirmation='False'|AcceptLicense='True'|
AllowUnofficialBuild='False'|Input='visualstudio2013ultimate'|
AllVersions='False'|SkipPackageInstallProvider='False'|
PackageNames='visualstudio2013ultimate'|Prerelease='False'|
ForceX86='False'|OverrideArguments='False'|NotSilent='False'|
IgnoreDependencies='False'|AllowMultipleVersions='False'|
ForceDependencies='False'|Information.PlatformType='Windows'|
Information.PlatformVersion='6.3.9600.0'|
Information.PlatformName='Windows Server 2012 R2'|
Information.ChocolateyVersion='0.9.9.8'|
Information.ChocolateyProductVersion='0.9.9.8'|
Information.FullName='choco, Version=0.9.9.8, Culture=neutral, PublicKeyToken=79d02ea9cad655eb'|

Information.Is64Bit='True'|Information.IsInteractive='False'|
Information.IsUserAdministrator='True'|
Information.IsProcessElevated='True'|Features.AutoUninstaller='False'|
Features.CheckSumFiles='True'|Features.FailOnAutoUninstaller='False'|
ListCommand.LocalOnly='False'|
ListCommand.IncludeRegistryPrograms='False'|
UpgradeCommand.FailOnUnfound='False'|
UpgradeCommand.FailOnNotInstalled='False'|
UpgradeCommand.NotifyOnlyAvailableUpgrades='False'|
NewCommand.AutomaticPackage='False'|SourceCommand.Command='unknown'|
FeatureCommand.Command='unknown'|PushCommand.TimeoutInSeconds='0'|
PinCommand.Command='unknown'|
2015-09-23 18:22:13,585 [DEBUG] - _ Chocolatey:ChocolateyInstallCommand - Normal Run Mode _
2015-09-23 18:22:13,585 [INFO ] - Installing the following packages:
2015-09-23 18:22:13,585 [INFO ] - visualstudio2013ultimate
2015-09-23 18:22:13,585 [INFO ] - By installing you accept licenses for the packages.
2015-09-23 18:22:17,538 [INFO ] - 
DotNet4.5.1 v4.5.1.20140606
2015-09-23 18:22:17,585 [DEBUG] - Contents of 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\Tools\ChocolateyInstall.ps1':
2015-09-23 18:22:17,600 [DEBUG] - if(-not (test-path "hklm:\SOFTWARE\Microsoft\.NETFramework\v4.0.30319\SKUs\.NETFramework,Version=v4.5.1")) {
Install-ChocolateyPackage 'dotnet451' 'exe' "/Passive /NoRestart /Log $env:temp\net451.log" 'http://download.microsoft.com/download/7/4/0/74078A56-A3A1-492D-BBA9-865684B83C1B/NDP451-KB2859818-Web.exe' -validExitCodes @(0,3010)
}
else {
     Write-Host "Microsoft .Net 4.5.1 Framework is already installed on your machine."
 } 

2015-09-23 18:22:17,616 [DEBUG] - Calling command ['"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -NoLogo -ExecutionPolicy Bypass -Command "[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\Tools\ChocolateyInstall.ps1' -installArguments '' -packageParameters ''"']
2015-09-23 18:22:18,491 [INFO ] -  Microsoft .Net 4.5.1 Framework is already installed on your machine.
2015-09-23 18:22:18,522 [DEBUG] - Command ['"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -NoLogo -ExecutionPolicy Bypass -Command "[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\Tools\ChocolateyInstall.ps1' -installArguments '' -packageParameters ''"'] exited with '0'
2015-09-23 18:22:18,522 [DEBUG] - Calling command ['"shutdown" /a']
2015-09-23 18:22:18,569 [DEBUG] - Command ['"shutdown" /a'] exited with '1116'
2015-09-23 18:22:18,585 [DEBUG] - Capturing package files in 'C:\ProgramData\chocolatey\lib\DotNet4.5.1'
2015-09-23 18:22:18,585 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\dotnet.png'
  with checksum '0D5A15B93F47DF7F91B24A14C7CE91FE'
2015-09-23 18:22:18,585 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\DotNet4.5.1.nupkg'
  with checksum '644E236231ADA16E07DA6C9B9612F508'
2015-09-23 18:22:18,585 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\DotNet4.5.1\Tools\ChocolateyInstall.ps1'
  with checksum '805DDE099F331DF8A4F0D47BE8D47E7C'
2015-09-23 18:22:18,600 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\extensions".
2015-09-23 18:22:18,600 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey".
2015-09-23 18:22:18,616 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey\DotNet4.5.1.4.5.1.20140606".
2015-09-23 18:22:18,631 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\DotNet4.5.1.4.5.1.20140606\.sxs".
2015-09-23 18:22:18,631 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\DotNet4.5.1.4.5.1.20140606\.pin".
2015-09-23 18:22:18,631 [INFO ] -  The install of dotnet4.5.1 was successful.
2015-09-23 18:22:18,631 [INFO ] - 
VisualStudio2013Ultimate v12.0.21005.20141031
2015-09-23 18:22:18,631 [DEBUG] - Contents of 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\ChocolateyInstall.ps1':
2015-09-23 18:22:18,631 [DEBUG] - . (Join-Path $(Split-Path -parent $MyInvocation.MyCommand.Definition) 'common.ps1')

$adminFile = (Join-Path $(Split-Path -parent $MyInvocation.MyCommand.Definition) 'AdminDeployment.xml')
$customArgs = $env:chocolateyInstallArguments
$env:chocolateyInstallArguments=""

$settings = Initialize-VS-Settings $customArgs $adminFile
$installerArgs = Get-VS-Installer-Args $settings.ProductKey

Install-ChocolateyPackage 'VisualStudio2013Ultimate' 'exe' $installerArgs 'http://download.microsoft.com/download/C/F/B/CFBB5FF1-0B27-42E0-8141-E4D6DA0B8B13/vs_ultimate_download.exe'

2015-09-23 18:22:18,631 [DEBUG] - Calling command ['"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -NoLogo -ExecutionPolicy Bypass -Command "[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\ChocolateyInstall.ps1' -installArguments '' -packageParameters ''"']
2015-09-23 18:22:19,928 [INFO ] -  Downloading VisualStudio2013Ultimate 32 bit
2015-09-23 18:22:19,944 [INFO ] -    from 'http://download.microsoft.com/download/C/F/B/CFBB5FF1-0B27-42E0-8141-E4D6DA0B8B13/vs_ultimate_download.exe'
2015-09-23 18:22:22,428 [INFO ] -  Installing VisualStudio2013Ultimate...
2015-09-23 18:22:29,038 [INFO ] -  VisualStudio2013Ultimate has been installed.
2015-09-23 18:22:29,053 [DEBUG] - Command ['"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -NoLogo -ExecutionPolicy Bypass -Command "[System.Threading.Thread]::CurrentThread.CurrentCulture = '';[System.Threading.Thread]::CurrentThread.CurrentUICulture = ''; & import-module -name 'C:\ProgramData\chocolatey\helpers\chocolateyInstaller.psm1'; & 'C:\ProgramData\chocolatey\helpers\chocolateyScriptRunner.ps1' -packageScript 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\ChocolateyInstall.ps1' -installArguments '' -packageParameters ''"'] exited with '0'
2015-09-23 18:22:29,053 [DEBUG] - Calling command ['"shutdown" /a']
2015-09-23 18:22:29,085 [DEBUG] - Command ['"shutdown" /a'] exited with '1116'
2015-09-23 18:22:29,100 [DEBUG] - Capturing package files in 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate'
2015-09-23 18:22:29,100 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\VisualStudio2013Ultimate.nupkg'
  with checksum '307AA81594B864085363287C08054360'
2015-09-23 18:22:29,100 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\AdminDeployment.xml'
  with checksum '01F61D4825533688114F2225072F92E4'
2015-09-23 18:22:29,100 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\ChocolateyInstall.ps1'
  with checksum '557ED94554A30A376F8C4580ADA76041'
2015-09-23 18:22:29,100 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\ChocolateyUninstall.ps1'
  with checksum 'F1A1396EC8B29639871B06C1360ACC3D'
2015-09-23 18:22:29,100 [DEBUG] -  Found 'C:\ProgramData\chocolatey\lib\VisualStudio2013Ultimate\Tools\common.ps1'
  with checksum 'CAD2FE3515E9D169FA58CD3ECC49A900'
2015-09-23 18:22:29,100 [DEBUG] - Attempting to create directory "C:\ProgramData\chocolatey\.chocolatey\VisualStudio2013Ultimate.12.0.21005.20141031".
2015-09-23 18:22:29,100 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\VisualStudio2013Ultimate.12.0.21005.20141031\.sxs".
2015-09-23 18:22:29,100 [DEBUG] - Attempting to delete file "C:\ProgramData\chocolatey\.chocolatey\VisualStudio2013Ultimate.12.0.21005.20141031\.pin".
2015-09-23 18:22:29,100 [INFO ] -  The install of visualstudio2013ultimate was successful.
```
If I manually run `choco install --force visualstudio2013ultimate` after all of this, it will install just fine.

Any suggestions?

Windows server 2012R2 with .NET 4.5.2 installed (manually) build from an AWS AMI

**Comment by svmastersamurai** [comment]
TL;DR: I think the installation exe needs to be run in interactive mode, which you can replicate in Chef by importing a scheduled task XML file that explicitly runs chocolatey as the local administrators group.

I had this same problem with VS 2015 Enterprise and had to come up with a somewhat unorthodox solution. My thesis is that the VS installation program needs to be run in interactive mode in order to work. When chocolatey is executed during a chef converge it is not running in an interactive session in our environment.

However, if you schedule a task that fires off the exact same chocolatey command it will work since you can set the task to run as an interactive user. The key for me was having the task run as a the administrators group, but you can't specify a runas group with schtasks on the command line. You can, however, dump the XML definition of said task and lay it down with a chef template! 
```
windows_task 'import_task_xml' do
  command "schtasks /create /xml '#{your_hacky_xml_path}' /tn vs_install"
  frequency :monthly
  run_level :highest
  action [:create, :run]
end

template your_hacky_xml_path do
  source 'install_visualstudio_task.xml.erb'
  action :create
end

powershell_script 'install visualstudio' do
  not_if { is_visualstudio_installed }
  code <<-EOF
schtasks /run /tn vs_install
do { Sleep 15 } while (ps -name choco)
do { Sleep 15 } while (ps -name vs_install_exe_thingy)
EOF
  action :run
end
```
I made the task in the task scheduler GUI and played with it until I got it into the state where it showed the chocolatey package downloading in a separate cmd window and the VS installation progress bar showed up, then did a right-click and export to the cookbook's file directory.
That last powershell_script resource I would say is optional if you are in charge of building the chocolatey package since you can bake the looping logic into the install script. In my case I was not, so I had to hack around it.
Might be an idea to get you unstuck ^_^

**Comment by mwrock** [comment]
Are you running the chef client remotely via winrm? The issue with dotnet and visual studio is that they pull down some bits via windows update server which will fail via winrm. Using a scheduled task will allow it to run in a local context. It's not that it needs an interactive user so much as a local one.

**Comment by svmastersamurai** [comment]
@mwrock Spot on! In my particular case I am triggering the installation via packer, which is using remote winrm.

**Comment by jedugas** [comment]
Same is true for SQL Server over WinRM, something I discovered recently with TestKitchen. There the root of the issue is a Crypto permissions problem, however it's defeated the same way (scheduled task  route, using a special provider the name of which is obvious but eludes me anyway!).

**Comment by svmastersamurai** [comment]
@jedugas Interesting, if you remember what that provider is let me know. I'd love to clean up my hackiness :)

**Comment by jedugas** [comment]
This is is, specific to TestKitchen...
https://github.com/smurawski/chef-zero-scheduled-task



On Thu, Dec 24, 2015 at 11:05 AM -0800, "Dan Sedlacek" <notifications@github.com> wrote:












@jedugas Interesting, if you remember what that provider is let me know. I'd love to clean up my hackiness :)

**Comment by svmastersamurai** [comment]
Thanks! :+1:

**Comment by mwrock** [comment]
closing now since it seems lie this is not an issue with the cookbook and workarounds have been found.
